File: HeapQueues/TaskScheduler.rs
use std::cmp::Reverse;
use std::collections::BinaryHeap;
use std::collections::VecDeque;

// this is a re-write of the above question in Rust
struct Solution {}

impl Solution {
    pub fn least_interval(tasks: Vec<char>, n: i32) -> i32 {
        let mut queue: VecDeque<(i32, i32)> = VecDeque::new();
        let mut freq: Vec<i32> = vec![0; 26];
        // create a frequency map
        for ch in &tasks {
          freq[ch-'A']++;
        }
        let mut max_heap: BinaryHeap<Reverse<i32>> = BinaryHeap::from(freq);
        let mut time = 0;

        while (!queue.is_empty() || !max_heap.is_empty()) {
          time++;
          if let mut Some(top) = max_heap.pop() {
            queue.push((top, time + n));
          } else {
            time = queue.front().2;
          }

          // now, pop all that are now free in this time
          while !queue.is_empty() && queue.front().unwrap().1 == time {
            max_heap.push(queue.pop().unwrap().1);
          }
        }
        time
    }
}

File: Cpp+Python/HeapQueues/TaskScheduler.cc
#include <ctime>
#include <queue>
#include <utility>
#include <vector>

using namespace std;

class Solution {
public:
  int leastInterval(vector<char> &tasks, int n) {
    // we actually don't really care about the character there
    priority_queue<int> pq;
    vector<int> freq(26, 0);
    queue<pair<int, int>> q;
    int time = 0;

    // populate the frequency map
    for (char &ch : tasks) {
      freq[ch - 'a']++;
    }

    // push those into the pq
    for (int i = 0; i < freq.size(); i++) {
      if (freq[i] != 0) {
        pq.push(freq[i]);
      }
    }

    while (!pq.empty() || !q.empty()) {
      time++;
      if (!pq.empty()) {
        int top = pq.top();
        if (top != 1) {
          q.push(make_pair(top - 1, time + n));
        }
        pq.pop();
      } else {
        time = q.front().second;
      }

      // now, all the tasks that can run again need to be popped, and
      // re-inserted into the heap
      while (!q.empty() && q.front().second == time) {
        pq.push(q.front().first);
        q.pop();
      }
    }
    return time;
  }
};
